crates/cards/src/validate.rs

```diff
@@ -0,0 +1,146 @@
+use std::{fmt, ops::Deref, str::FromStr};
+
+use masking::{PeekInterface, Strategy, StrongSecret, WithType};
+use serde::{Deserialize, Deserializer, Serialize};
+use thiserror::Error;
+
+#[derive(Debug, Deserialize, Serialize, Error)]
+#[error("not a valid credit card number")]
+pub struct CCValError;
+
+impl From<core::convert::Infallible> for CCValError {
+    fn from(_: core::convert::Infallible) -> Self {
+        Self
+    }
+}
+
+/// Card number
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize)]
+pub struct CardNumber(StrongSecret<String, CardNumberStrategy>);
+
+impl FromStr for CardNumber {
+    type Err = CCValError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match luhn::valid(s) {
+            true => Ok(Self(StrongSecret::from_str(s)?)),
+            false => Err(CCValError),
+        }
+    }
+}
+
+impl TryFrom<String> for CardNumber {
+    type Error = CCValError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::from_str(&value)
+    }
+}
+
+impl Deref for CardNumber {
+    type Target = StrongSecret<String, CardNumberStrategy>;
+
+    fn deref(&self) -> &StrongSecret<String, CardNumberStrategy> {
+        &self.0
+    }
+}
+
+impl<'de> Deserialize<'de> for CardNumber {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        let s = String::deserialize(d)?;
+        Self::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
+
+impl CardNumber {
+    pub fn remove_whitespace(self) -> Self {
+        let cc_no_whitespace: String = self.peek().split_whitespace().collect();
+        let secret = StrongSecret::new(cc_no_whitespace);
+        Self(secret)
+    }
+}
+
+pub struct CardNumberStrategy;
```

Comment:
You should be able to mark this private, since no types outside this module would need it.

```suggestion
struct CardNumberStrategy;
```

Reply:
The problem is that I use it in the `Deref` trait which is a public interface and hence I have to make `CardNumberStrategy` public as well.
If I understood everything correctly.

Reply:
Alright then you can ignore my comment. Marking this conversation resolved.